File: frontend/src/types/cogs.ts
/**
 * ARCH-18 / ARCH-24 — COGS, unit economics and supplier reconciliation DTOs.
 *
 * Every `| null` in this file is load-bearing. The backend returns null for a
 * margin it cannot compute, a share it cannot take, and a ratio that is
 * undefined — and the UI must render those as "unknown" rather than as zero.
 * Widening any of these to a plain `number` would let a `?? 0` slip in at the
 * call site and turn "we don't know what this cost" into "this was free",
 * which reads on the dashboard as a 100% gross margin.
 *
 * Per-unit prices arrive as strings because they carry nine decimal places
 * and would lose precision through a JSON float. They are display-only; every
 * total is an integer micros field.
 */

/* -------------------------------------------------------------------------
 * Margins
 * ---------------------------------------------------------------------- */

export interface MarginFigures {
  readonly revenue_micros: number;
  /** Revenue on rows that also carry a cost basis — the margin's denominator. */
  readonly attributed_revenue_micros: number;
  readonly cost_basis_micros: number;
  readonly unknown_cost_revenue_micros: number;

  /** null when no row in the window has a known cost. */
  readonly gross_margin_micros: number | null;
  readonly gross_margin_ratio: number | null;

  /** Share of revenue excluded from the margin, by value. Read this first. */
  readonly unknown_cost_share: number | null;
  /** Share of attributed revenue whose cost is ESTIMATED rather than measured. */
  readonly soft_cost_share: number | null;

  readonly event_count: number;
  readonly known_cost_event_count: number;
  readonly unknown_cost_event_count: number;

  /** False when too little revenue has a known cost for the margin to be quoted. */
  readonly is_trustworthy: boolean;
}

export interface PlatformMarginSummary {
  readonly period_start: string;
  readonly period_end: string;
  readonly currency: string;
  readonly organization_count: number;
  readonly figures: MarginFigures;
}

export interface TenantEconomicsEntry {
  readonly organization_id: string;
  readonly organization_name: string | null;
  readonly organization_slug: string | null;
  readonly figures: MarginFigures;
}

export type MarginOrder =
  | "MARGIN_ASC"
  | "MARGIN_DESC"
  | "REVENUE_DESC"
  | "UNKNOWN_DESC";

export interface TenantEconomicsResponse {
  readonly period_start: string;
  readonly period_end: string;
  readonly currency: string;
  readonly order: MarginOrder;
  readonly entries: readonly TenantEconomicsEntry[];
}

export interface ProviderCostEntry {
  readonly provider: string | null;
  readonly cost_basis_micros: number;
  readonly revenue_micros: number;
  readonly event_count: number;
  readonly unknown_cost_event_count: number;
}

export interface ProviderCostResponse {
  readonly period_start: string;
  readonly period_end: string;
  readonly entries: readonly ProviderCostEntry[];
}

/* -------------------------------------------------------------------------
 * Rate card
 * ---------------------------------------------------------------------- */

export type CostBasisSource =
  | "SUPPLIER_RATE_CARD"
  | "MEASURED"
  | "ESTIMATED"
  | "ZERO_BYOK";

export interface RateCardEntry {
  readonly event_type: string;
  readonly provider: string;
  readonly model: string | null;
  readonly tier_key: string | null;
  readonly unit: string;
  readonly unit_price_micros: string;
  readonly cost_basis_micros: string | null;
  readonly cost_basis_source: CostBasisSource | null;
  readonly unit_margin_micros: string | null;
  readonly notes: string | null;
}

export interface RateCardResponse {
  readonly price_book_id: string | null;
  readonly price_book_version: number | null;
  readonly currency: string | null;
  readonly effective_from: string | null;
  readonly entry_count: number;
  readonly with_cost_basis: number;
  readonly hard_cost_basis: number;
  readonly coverage_ratio: number | null;
  readonly entries: readonly RateCardEntry[];
}

/* -------------------------------------------------------------------------
 * Supplier invoices
 * ---------------------------------------------------------------------- */

export type ReconciliationStatus = "MATCHED" | "INVESTIGATE" | "ACCEPTED";

/**
 * ARCH-24. Which denominator produced a variance figure.
 *
 * `ARCH14_SELL_SIDE` is denominated in customer price and is inflated by our
 * own gross margin. It is NOT cost variance and must never be rendered as
 * COGS, however tempting the adjacency.
 */
export type CostBasisMethod =
  | "ARCH18_SUPPLIER_COST"
  | "ARCH18_PRE_CONSOLIDATION"
  | "ARCH14_SELL_SIDE";

/** ARCH-24 N-3. Where a supplier invoice row came from. */
export type SupplierInvoiceOrigin = "STATEMENT_PULL" | "OPERATOR_UPLOAD";

export interface SupplierReconciliation {
  readonly id: string;
  readonly supplier_invoice_id: string;
  readonly modelled_total_micros: number;
  readonly variance_micros: number;
  /** null when the modelled total is zero — undefined, not a perfect match. */
  readonly variance_ratio: number | null;
  readonly status: ReconciliationStatus;
  readonly modelled_event_count: number;
  readonly unknown_cost_event_count: number;
  /** ARCH-24: Denominator method indicator. */
  readonly cost_basis_method: CostBasisMethod;
  /** ARCH-24: True when variance is denominated in genuine supplier cost. */
  readonly is_authoritative_cost: boolean;
  readonly note: string | null;
  readonly reconciled_at: string;
  readonly reconciled_by_user_id: string | null;
}

export interface SupplierInvoice {
  readonly id: string;
  readonly provider: string;
  readonly invoice_reference: string | null;
  readonly period_start: string;
  /** Last day covered, INCLUSIVE. */
  readonly period_end: string;
  readonly invoiced_total_micros: number;
  readonly currency: string;
  readonly raw_document_file_id: string | null;
  readonly ingested_at: string;
  readonly ingested_by_user_id: string | null;
  readonly notes: string | null;
  readonly latest_reconciliation: SupplierReconciliation | null;
  readonly origin: SupplierInvoiceOrigin;
  readonly superseded_invoice_id: string | null;
}

export interface ConsolidatedReconciliationResponse {
  readonly entries: readonly SupplierReconciliation[];
  readonly authoritative_method: CostBasisMethod;
  /**
   * ARCH-14 runs in the same window. A count only, deliberately: their drift
   * figures are customer-price denominated and showing them beside these rows
   * would invite exactly the misreading ARCH-24 removed.
   */
  readonly sell_side_run_count: number;
}

export interface RollupCostBasisEntry {
  readonly organization_id: string;
  readonly granularity: string;
  readonly bucket_start: string;
  readonly event_type: string;
  readonly provider: string | null;
  readonly event_count: number;
  readonly cost_micros: number;
  /**
   * NULL means unknown, not zero. Buckets sealed before ARCH-24 are
   * permanently null by design. Render "unknown" — never a dash that reads
   * as free, and never a margin computed against it.
   */
  readonly cost_basis_micros: number | null;
  readonly unknown_cost_basis_event_count: number;
  readonly cost_basis_source_mix: Readonly<Record<string, number>> | null;
  /** Computed by the backend. The client never recomputes the threshold. */
  readonly is_trustworthy: boolean;
}

export interface SupplierInvoiceListResponse {
  readonly entries: readonly SupplierInvoice[];
}

export interface SupplierInvoiceCreateRequest {
  readonly provider: string;
  readonly period_start: string;
  readonly period_end: string;
  readonly invoiced_total_micros: number;
  readonly currency?: string;
  readonly invoice_reference?: string | null;
  readonly notes?: string | null;
}

export interface ReconcileRequest {
  readonly note?: string | null;
  readonly threshold_ratio?: number | null;
  readonly force?: boolean;
}

export interface AcceptVarianceRequest {
  readonly note: string;
}

/* -------------------------------------------------------------------------
 * Formatting
 * ---------------------------------------------------------------------- */

export const UNKNOWN_LABEL = "unknown";

/** Micros to a currency string. 1_000_000 micros = 1 unit. */
export const formatMicros = (
  micros: number | null | undefined,
  currency = "USD",
): string => {
  if (micros === null || micros === undefined) {
    return UNKNOWN_LABEL;
  }
  const units = micros / 1_000_000;
  try {
    return new Intl.NumberFormat(undefined, {
      style: "currency",
      currency,
      maximumFractionDigits: Math.abs(units) < 1 ? 4 : 2,
    }).format(units);
  } catch {
    return `${units.toFixed(2)} ${currency}`;
  }
};

export const formatRatio = (ratio: number | null | undefined): string => {
  if (ratio === null || ratio === undefined || !Number.isFinite(ratio)) {
    return UNKNOWN_LABEL;
  }
  return `${(ratio * 100).toFixed(1)}%`;
};

export const formatSignedMicros = (
  micros: number | null | undefined,
  currency = "USD",
): string => {
  if (micros === null || micros === undefined) {
    return UNKNOWN_LABEL;
  }
  const sign = micros > 0 ? "+" : "";
  return `${sign}${formatMicros(micros, currency)}`;
};

/**
 * How much of a margin figure to believe.
 *
 * Mirrors margin_service.MIN_TRUSTWORTHY_KNOWN_SHARE, but the UI never
 * recomputes the verdict — the backend sends `is_trustworthy` and this only
 * chooses the wording.
 */
export const confidenceLabel = (figures: MarginFigures): string => {
  if (figures.event_count === 0) {
    return "No usage in this period";
  }
  if (figures.gross_margin_micros === null) {
    return "No cost basis recorded for any usage in this period";
  }
  if (!figures.is_trustworthy) {
    return `Margin excludes ${formatRatio(figures.unknown_cost_share)} of revenue`;
  }
  if ((figures.soft_cost_share ?? 0) > 0.2) {
    return `${formatRatio(figures.soft_cost_share)} of cost is estimated`;
  }
  return "Cost basis known for most revenue";
};

export const statusTone = (
  status: ReconciliationStatus,
): "ok" | "warn" | "muted" => {
  switch (status) {
    case "MATCHED":
      return "ok";
    case "INVESTIGATE":
      return "warn";
    case "ACCEPTED":
    default:
      return "muted";
  }
};

File: backend/app/schemas/cogs.py
"""ARCH-18 — COGS, margin and supplier reconciliation DTOs.

Every optional numeric field here is optional on purpose. `gross_margin_micros`
is None when no cost is known; `unknown_cost_share` is None when there is no
revenue to take a share of; `variance_ratio` is None when the modelled total is
zero. The frontend renders each of those as "unknown" and never as a number,
which is the whole reason they are not defaulted to 0 for the convenience of
the serializer.

Micros are serialised as JSON numbers, not strings. They are integers bounded
by BIGINT, and JavaScript holds integers exactly to 2^53 — about 9 billion
dollars expressed in micros. Ratios are floats because they are already
approximations and pretending otherwise in the wire format would be theatre.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_serializer

# ---------------------------------------------------------------------------
# Margins
# ---------------------------------------------------------------------------


class MarginFiguresResponse(BaseModel):
    """The numeric core shared by the platform summary and every tenant row."""

    revenue_micros: int
    attributed_revenue_micros: int = Field(
        description=(
            "Revenue on rows that also carry a cost basis. The only revenue a "
            "gross margin is computed against."
        )
    )
    cost_basis_micros: int
    unknown_cost_revenue_micros: int

    gross_margin_micros: Optional[int] = Field(
        default=None,
        description="None when no row in the window has a known cost.",
    )
    gross_margin_ratio: Optional[float] = None

    unknown_cost_share: Optional[float] = Field(
        default=None,
        description=(
            "Share of revenue excluded from the margin, by value not by row "
            "count. Read this before reading the margin."
        ),
    )
    soft_cost_share: Optional[float] = Field(
        default=None,
        description="Share of attributed revenue whose cost is ESTIMATED.",
    )

    event_count: int
    known_cost_event_count: int
    unknown_cost_event_count: int

    is_trustworthy: bool = Field(
        description=(
            "False when too little of the revenue has a known cost for the "
            "margin to be quotable. The dashboard suppresses the headline "
            "figure rather than showing a confident wrong number."
        )
    )

    model_config = ConfigDict(from_attributes=True)


class PlatformMarginSummaryResponse(BaseModel):
    period_start: datetime
    period_end: datetime
    currency: str
    organization_count: int
    figures: MarginFiguresResponse

    model_config = ConfigDict(from_attributes=True)


class TenantEconomicsEntry(BaseModel):
    organization_id: uuid.UUID
    organization_name: Optional[str] = None
    organization_slug: Optional[str] = None
    figures: MarginFiguresResponse

    model_config = ConfigDict(from_attributes=True)


class TenantEconomicsResponse(BaseModel):
    period_start: datetime
    period_end: datetime
    currency: str
    order: str
    entries: list[TenantEconomicsEntry] = []

    model_config = ConfigDict(from_attributes=True)


class ProviderCostEntry(BaseModel):
    provider: Optional[str] = None
    cost_basis_micros: int
    revenue_micros: int
    event_count: int
    unknown_cost_event_count: int

    model_config = ConfigDict(from_attributes=True)


class ProviderCostResponse(BaseModel):
    period_start: datetime
    period_end: datetime
    entries: list[ProviderCostEntry] = []

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Rate card
# ---------------------------------------------------------------------------


class RateCardEntry(BaseModel):
    event_type: str
    provider: str
    model: Optional[str] = None
    tier_key: Optional[str] = None
    unit: str
    unit_price_micros: Decimal
    cost_basis_micros: Optional[Decimal] = None
    cost_basis_source: Optional[str] = None
    unit_margin_micros: Optional[Decimal] = Field(
        default=None,
        description="price - cost. None when the cost is unknown.",
    )
    notes: Optional[str] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    @field_serializer("unit_price_micros", "cost_basis_micros", "unit_margin_micros")
    def _decimal_as_string(self, value: Optional[Decimal]) -> Optional[str]:
        # Nine decimal places of per-unit price does not survive a float. The
        # frontend treats these as display strings and never does arithmetic
        # on them; totals come from the integer micros fields above.
        return None if value is None else format(value, "f")


class RateCardResponse(BaseModel):
    price_book_id: Optional[uuid.UUID] = None
    price_book_version: Optional[int] = None
    currency: Optional[str] = None
    effective_from: Optional[datetime] = None
    entry_count: int = 0
    with_cost_basis: int = 0
    hard_cost_basis: int = 0
    coverage_ratio: Optional[float] = None
    entries: list[RateCardEntry] = []

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Supplier invoices and reconciliation
# ---------------------------------------------------------------------------


class SupplierInvoiceCreate(BaseModel):
    provider: str = Field(min_length=1, max_length=32)
    period_start: date
    period_end: date = Field(
        description="Last day covered, INCLUSIVE. '1 Jul - 31 Jul' is 31 Jul."
    )
    invoiced_total_micros: int = Field(ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    invoice_reference: Optional[str] = Field(default=None, max_length=200)
    raw_document_file_id: Optional[uuid.UUID] = Field(
        default=None,
        description=(
            "A platform-scoped upload only. A tenant's file is refused: the "
            "FK is RESTRICT and would pin that tenant's document against "
            "their own deletion."
        ),
    )
    notes: Optional[str] = Field(default=None, max_length=1000)

    model_config = ConfigDict(protected_namespaces=())


class SupplierReconciliationResponse(BaseModel):
    id: uuid.UUID
    supplier_invoice_id: uuid.UUID
    modelled_total_micros: int
    variance_micros: int
    variance_ratio: Optional[float] = None
    status: str
    modelled_event_count: int
    unknown_cost_event_count: int
    note: Optional[str] = None
    reconciled_at: datetime
    reconciled_by_user_id: Optional[uuid.UUID] = None

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("variance_ratio")
    def _ratio_as_float(self, value: Optional[float]) -> Optional[float]:
        return None if value is None else float(value)


class SupplierInvoiceResponse(BaseModel):
    id: uuid.UUID
    provider: str
    invoice_reference: Optional[str] = None
    period_start: date
    period_end: date
    invoiced_total_micros: int
    currency: str
    raw_document_file_id: Optional[uuid.UUID] = None
    ingested_at: datetime
    ingested_by_user_id: Optional[uuid.UUID] = None
    notes: Optional[str] = None
    latest_reconciliation: Optional[SupplierReconciliationResponse] = None

    model_config = ConfigDict(from_attributes=True)


class SupplierInvoiceListResponse(BaseModel):
    entries: list[SupplierInvoiceResponse] = []

    model_config = ConfigDict(from_attributes=True)


class ReconcileRequest(BaseModel):
    note: Optional[str] = Field(default=None, max_length=4000)
    threshold_ratio: Optional[float] = Field(
        default=None,
        ge=0.0,
        le=10.0,
        description="Override the configured variance threshold for this run.",
    )
    force: bool = Field(
        default=False,
        description=(
            "Reconcile a period that has not been closed long enough to be "
            "final. Recorded on the resulting row."
        ),
    )


class AcceptVarianceRequest(BaseModel):
    note: str = Field(
        min_length=1,
        max_length=4000,
        description=(
            "Mandatory. An accepted variance with no stated reason is "
            "indistinguishable from a mistake."
        ),
    )


MarginOrder = Literal["MARGIN_ASC", "MARGIN_DESC", "REVENUE_DESC", "UNKNOWN_DESC"]


__all__ = [
    "AcceptVarianceRequest",
    "MarginFiguresResponse",
    "MarginOrder",
    "PlatformMarginSummaryResponse",
    "ProviderCostEntry",
    "ProviderCostResponse",
    "RateCardEntry",
    "RateCardResponse",
    "ReconcileRequest",
    "SupplierInvoiceCreate",
    "SupplierInvoiceListResponse",
    "SupplierInvoiceResponse",
    "SupplierReconciliationResponse",
    "TenantEconomicsEntry",
    "TenantEconomicsResponse",
]